src/background/services/tabState.ts

```diff
@@ -0,0 +1,62 @@
+import { getHash } from '../utils'
+
+import { Tabs } from 'webextension-polyfill'
+import { WalletAddress } from '@interledger/open-payments'
+
+type State = {
+  lastPaymentTimestamp: number
+  expiresAtTimestamp: number
+}
+
+export class TabState {
+  private state = new WeakMap<Tabs.Tab, { [key: string]: State }>()
+
+  constructor() {}
+
+  private async getStateKey(
+    url: string,
+    walletAddress: WalletAddress
+  ): Promise<string> {
+    const hashUrl = await getHash(url)
+    const hashWalletAddress = await getHash(JSON.stringify(walletAddress))
+    const key = `${hashUrl}:${hashWalletAddress}`
```

Comment:
Another: do we need URL in here? Tab intrinsically has the URL, and we mostly care about not paying same walletAddress in same tab here.

Reply:
we can have the same walletAddress in 2 different iframes, so we need the url

Reply:
So, if a site adds 10 iframes (with different URLs) using same wallet address on a single page, that won't count as overpaying?

Reply:
I would say no. We can discuss with the team and treat case this in another task